docs/source/blueprints/BP005.rst

```diff
@@ -0,0 +1,291 @@
+either be described (documentation) and enforced as automated checks as much as
+possible.
+
+Among the topics related to the quality requirements and standards, the
+contribution documentation and automated checks will cover, at least:
+
+* Logical, stylistic, analytical and formatter checks
+
+  * Docstring checks
+  * Pylint
+  * PEP-8, PEP-257, etc…
+
+* Unit and Functional Tests Coverage
+
+  Due to the current code quality and given this is a new repository we would
+  like to reach at least 90% of coverage. This way, users writing tests relying
+  on these libraries will have a guarantee of a good level of correctness and
+  quality, redirecting the debugging and troubleshooting efforts away from the
+  well tested libraries.
+
+The new standalone repository will have predefined pipelines to run those
+checks. Newly added libraries will automatically be subject to those checks.
+
+Metadata description
+--------------------
+
+Every module will need to provide a description including:
+
+* Utility description
+* The utility category (ie: network, cpu, security, etc..)
+* Maintainers information
+  * Main contact (We need a contact point, a single maintainer or a group)
+  * Maintainer Group members (this will be used to block merges)
+* Supported Platforms
+* Supported execution method (Local vs Remote)
+  We intend to flag modules that have the capability of acting on remote
+  systems. This will probably be a broad classification, just for informational
+  purposes because unfortunately we have way too many ways of executing code
+  remotely.
+
+  We don't hope to address a complete standardization of the remote execution
+  APIs just yet, but we want to at least bring some visibility to it.
+
+For instance, the textual category and description will allow users to search
+for modules based on their interest.  Also, our pipelines will parse that data
+field containing the list of supported platforms, and will trigger tests on
+those platforms. Users will then be able to see the results of those tests on
+each of the platforms defined, giving them quality assurance and coverage
+there.
+
+The exact format of the metadata file is still to be defined, but a possibility
+is to define a JSON schema that will be provided on this new repository.
+
+Maintainability
+===============
+
+As described above, each library must follow the new standard and keep that
+same level for future versions. Because of that, new PRs must be approved by at
+least one MAINTAINER and pass the basic checks pipelines that will be provided
+as part of the autils repository.
+
+Besides the automatic checks, each MAINTAINER is responsible for keeping the
+consistency of that library with the overall project.
+
+Release
+=======
+
+As a new and standalone repository, there will not be any type of release
+version sync between Avocado and autils.  Avocado may release, say, version
+100.0, while autils won’t have a release. Because of that, autils will be
```

Comment:
Here one question is: would avocado consume autils?

If yes, another question is what are the rules for avocado's plugins (for instance, avocado-vt) to consume this repo as well? Let's suppose that avocado 100.0 would have the requirement of this repo as `autils >= 5.0`, then can avocado-vt also give its own requirement (say, a different one with avocado's) to the same repo?

Reply:
> Here one question is: would avocado consume autils?
> 

I would expect that to be a yes.  There's a small, but maybe still meaningful, set of "avocado.utils" libraries which are used by Avocado core that *could*, with some effort, be kept as part of the Avocado repo.  But, there's a larger set of modules that are used by both Avocado core and "avocado.utils".  So, IMO, it makes sense to have all generally useful modules moved to "avocado.utils", and thus, Avocado (core) using autils.

> If yes, another question is what are the rules for avocado's plugins (for instance, avocado-vt) to consume this repo as well? Let's suppose that avocado 100.0 would have the requirement of this repo as `autils >= 5.0`, then can avocado-vt also give its own requirement (say, a different one with avocado's) to the same repo?

This is more of a Python/packaging issue than an Avocado/Avocado-VT/autils policy.  The Python solution for multiple version requirement is to have each of these combinations in a different venv.  This, of course, doesn't work when you have a layered requirement relationship (both Avocado and Avocado-VT must be on the same environment, be it a venv or not).

So, my expectation is that we would expand the idea of planning  Avocado versions known to work with Avocado-VT versions (instead of trying to guarantee compatibility at every single commit).  One of these planning points should be sync the usage of autils to the same version.

I expect that Avocado-VT, because of the requirements of newly added Avocado-VT tests (say from tp-qemu, tp-libvirt), would drive the upgrading of autils version.  Avocado (core) would then need to be adjusted to keep up the new autils changes.  I expect that the required changes on Avocado's side would be rather small, given the also small set of libraries used by Avocado core.

Reply:
Thanks @clebergnu for clarifying.

I'm resolving this conversation so we can approve this Blueprint. Please reopen a new thread if you would like to get something fixed on this draft.